storage/gcloud/aio/storage/storage.py

```diff
@@ -485,18 +490,58 @@ def _get_stream_len(stream: IO[AnyStr]) -> int:
             stream.seek(current)
 
     @staticmethod
-    def _preprocess_data(data: Any) -> IO[Any]:
+    def _compress_file_in_chunks(input_stream: Union[IO[AnyStr], io.IOBase],
+                                 output_stream: io.IOBase,
+                                 chunk_size: int = 8192) -> None:
+        """
+        Reads the contents of input_stream and writes it gzip-compressed to
+        output_stream in chunks. The chunk size is 8Kb by default, which is a
+        standard filesystem block size.
+        """
+        with gzip.open(output_stream, 'wb') as f_out:
+            while True:
+                chunk = input_stream.read(chunk_size)
+                if not chunk:
+                    break
+                if isinstance(chunk, str):
+                    chunk_bytes = chunk.encode('utf-8')
+                else:
+                    chunk_bytes = chunk
+                # At this point we assume that chunk is a bytes object
+                f_out.write(chunk_bytes)
+
+        # After finishing writing, reset the buffer position,
+        # so it can be read
+        output_stream.seek(0)
+
+    @staticmethod
+    def _preprocess_data(
+            data: Any,
+            gzip_compress: bool = False,
+    ) -> IO[Any]:
+
+        stream: Union[IO[Any], io.IOBase]
         if data is None:
-            return io.StringIO('')
+            stream = io.StringIO('')
+        elif isinstance(data, bytes):
+            stream = io.BytesIO(data)
+        elif isinstance(data, str):
+            stream = io.StringIO(data)
+        elif isinstance(data, io.IOBase):
+            stream = data
+        else:
+            raise TypeError(f'unsupported upload type: "{type(data)}"')
 
-        if isinstance(data, bytes):
-            return io.BytesIO(data)
-        if isinstance(data, str):
-            return io.StringIO(data)
-        if isinstance(data, io.IOBase):
-            return data  # type: ignore[return-value]
+        if gzip_compress:
```

Comment:
You should be able to keep the type hint breakages simpler by doing this as a second pass: eg. instead of doing this block within `_preprocess_data()`, we can just do it as a second step in `upload()` and avoid modifying this function at all.

So instead of how you have:
```
stream = self._preprocess_data(file_data, gzip_compress=zipped)
```
We can do:
```
stream = self._preprocess_data(file_data)
if zipped:
    stream = self._compress_file_in_chunks(stream)
```